Remove support for specifying nvidia driver capabilities in --gpus flag

elezar opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Updated the code that handles the injection of the `nvidia-container-runtime-hook` to not handle NVIDIA-specific capabilities instead relying on the `NVIDIA_DRIVER_CAPABILITIES` environment variable.

The motivation for this is:
1. The list of capabilities is incomplete and has not been updated for new capabilities.
2. Handling of capabilities complicats the migration to CDI
3. The handling of capabilities on the --gpus flag conflicts with the NVIDIA_DRIVER_CAPABILITES envvar that is specified in the image or on the command line.

See also:
- #49952
- #49824

**- How I did it**

Removed references to and processing of NVIDIA-specific capabilities.

**- How to verify it**

Running: 
```
docker run --rm -ti --gpus `all,capabilities=utility` -e NVIDIA_DRIVER_CAPABILTIES=all ubuntu env | grep NVIDIA_DRIVER_CAPABILTIES
```
should show:
```
NVIDIA_DRIVER_CAPABILITIES=all
``` 

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
Removed support for specifying NVIDIA-specific capabilities. This affects the `--gpus` flag and the Docker Compose specification.
```

**- A picture of a cute animal (not mandatory but encouraged)**

elezar commented:
cc @thaJeztah

thaJeztah commented:
Curious; would these options be something we should produce an error for when set, or (I guess they're a bit of an niche case) is it safe to silently ignore if they're used? Alternatively, something in the middle would be to return a warning;
https://github.com/moby/moby/blob/9663b36b6ddbb365c99546b7ae9bc8b51837c593/api/server/router/container/container_routes.go#L665-L668
https://github.com/moby/moby/blob/9663b36b6ddbb365c99546b7ae9bc8b51837c593/api/server/router/container/container_routes.go#L706-L707


> We would also need to remove the reference here: https://docs.docker.com/engine/containers/resource_constraints/#set-nvidia-capabilities

Yes, also the compose docs, looks like;
https://docs.docker.com/compose/how-tos/gpu-support/#example-of-a-compose-file-for-running-a-service-with-access-to-1-gpu-device


IIUC, those examples could be updated by setting the env-var manually instead, correct? If that's the case, and if that already works, perhaps would be good to already update the examples to match that.

> We would also need to adjust the compose spec:

Yup; not sure how deprecating works there (ISTR the compose-spec may not deprecate things, but implementations are allowed to return an error if they don't support a feature if set, so ... maybe it's just docs; not 100% sure though!)

We should probably a deprecation in;

- https://github.com/docker/cli/blob/master/docs/deprecated.md
- and a mention in the API changes (even though it's not modifying the API definition, it's removing options, so ... perhaps worth a mention? Not sure if we documented any of those capabilities though 🤔)

elezar committed:
Remove support for specifying nvidia driver capabilities in --gpus flag

This change removes support for specifying NVIDIA_DRIVER_CAPABILITIES as
capabilities in the --gpus flag (or through docker compose).

The motivation for this is:
1. The list of capabilities is incomplete and has not been
   updated for new capabilities.
2. Handling of capabilities complicats the migration to CDI
3. The handling of capabilities on the --gpus flag conflicts
   with the NVIDIA_DRIVER_CAPABILITES envvar that is specified
   in the image or on the command line.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

elezar committed:
Prefer requested driver over capabilities

This change ignores requested capabilities when a driver is explicitly
requested. This simplifies the logic for selecting a driver and means
that users need not spefify redundant capabilities.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

elezar commented on api/swagger.yaml:
```diff
@@ -315,8 +315,8 @@ definitions:
         example:
-          # gpu AND nvidia AND compute
-          - ["gpu", "nvidia", "compute"]
+          # gpu AND nvidia
+          - ["gpu", "nvidia"]
```
Given the ambiguity, is removing support for capabilities entirely (even `gpu & (nvidia | amd)`) and only supporting the driver name an option? I'm not sure what the community things about this as a breaking change.

elezar commented:
I have moved the non-deprecation changes to #50717.
